src/pipeline.rs

```diff
@@ -21,66 +30,137 @@ pub fn build_graphs(
+                                    let row = parse_tsv_line(line);
+                                    let line_col_num = row.len();
+                                    if line_col_num == config_col_num {
+                                        row_s.send(row).unwrap();
+                                    } else {
+                                        warn!("Wrong number of columns (expected: {}, provided: {}). The line [{}] is skipped.", config_col_num, line_col_num, line);
+                                    }
+                                });
+                            }
+                        });
                     }
-                });
+                }
             }
         }
-    }
 
-    drop(bus);
+        #[allow(clippy::needless_collect)]  // Must collect because of 'drop' calls below
+        let buffers_handles: Vec<_> = (0..processing_worker_num).map(|_| {
+            let hyperedges_r = hyperedges_r.clone();
+            let entity_processor = EntityProcessor::new(
+                config,
+                in_memory_entity_mapping_persistor.clone(),
+            );
+            let sparse_matrices = sparse_matrices_desc.clone();
 
-    let mut sparse_matrices = vec![];
-    for join_handle in sparse_matrix_threads {
-        let sparse_matrix = join_handle
-            .join()
-            .expect("Couldn't join on the associated thread");
-        sparse_matrices.push(sparse_matrix);
+            s.spawn(move |_| {
+                let mut buffers: Vec<_> = sparse_matrices.iter().map(|smd| smd.make_buffer()).collect();
+
+                for row in hyperedges_r {
+                    let iterator = entity_processor.process_row_and_get_edges(&row);
+
+                    for buffer in &mut buffers {
+                        for hashes in iterator.clone() {
+                            buffer.handle_pair(&hashes);
+                        }
+                    }
+                }
+
+                buffers
+            })
+        }).collect();
+
+        // Drop it so all channels are closed when work is finished
+        drop(hyperedges_s);
+        drop(hyperedges_r);
+
+        buffers_handles.into_iter().map(|h| h.join().expect("Thread finished job")).collect()
+    }).expect("Threads finished work");
+
+    let merging_time = Instant::now();
+
+    // (WORKER_IX, MATRIX_IX) -> (MATRIX_IX, WORKER_IX)
+    let buffers = transpose(buffers);
+
+    let result = buffers
+        .into_par_iter()
+        .zip(sparse_matrices_desc.par_iter())
+        .map(|(parts, desc)| {
+            assert_eq!(parts.len(), processing_worker_num);
+            SparseMatrixBuffer::finish(desc, parts)
+        })
+        .collect();
+
+    info!(
+        "Merging finished in {} sec",
+        merging_time.elapsed().as_secs()
+    );
+
+    result
+}
+
+fn transpose<T>(vec_2d: Vec<Vec<T>>) -> Vec<Vec<T>> {
+    let dim_0_size = vec_2d.len();
+    if dim_0_size == 0 {
+        panic!("Cannot transpose empty list. Dim=0")
```

Comment:
why panic? maybe just return vec_2d

Reply:
Not sure. In numpy for example it's legal to have shape <0, 5> thats getting transposed to <5, 0>

Since its a code branch that we control and that is incorrect in our context I like explicit loud panic more.

Reply:
Though you can have only shape (0,0) here, which I'd say is fine to transpose (and it works in numpy as well)

BTW.
```
>>> np.zeros((5,0)).T
array([], shape=(0, 5), dtype=float64)
```

Reply:
>Also you probably don't need to take desc as you can take one from buffers?

Removed

Reply:
>https://github.com/Synerise/cleora/pull/65#discussion_r923145134

Done

Reply:
On the other hand "0" here is imposible, so it's ok to panic. Something must have gone wrong if we get 0 buffers.